File: submission/neha/scripts/validate_schema.py
"""Validate that the database schema contains required tables and columns."""

from __future__ import annotations

import sqlite3
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))

from service.app import init_schema  # noqa: E402

REQUIRED: dict[str, set[str]] = {
    "wallets": {"id", "balance", "currency"},
    "transfers": {
        "id",
        "source_wallet_id",
        "destination_wallet_id",
        "amount",
        "currency",
        "status",
        "created_at",
    },
    "idempotency_keys": {"key", "payload_hash", "transfer_id", "created_at"},
    "audit_events": {"id", "transfer_id", "event_type", "created_at"},
    "outbox_events": {
        "id",
        "transfer_id",
        "event_type",
        "payload",
        "status",
        "created_at",
    },
}


def main() -> None:
    conn = sqlite3.connect(":memory:")
    init_schema(conn)

    errors: list[str] = []
    for table, required_cols in REQUIRED.items():
        rows = conn.execute(f"PRAGMA table_info({table})").fetchall()  # nosec B608
        existing = {row[1] for row in rows}
        missing = required_cols - existing
        if missing:
            errors.append(f"Table '{table}' missing columns: {missing}")
        else:
            print(f"OK: {table}")

    conn.close()

    if errors:
        for err in errors:
            print(f"ERROR: {err}", file=sys.stderr)
        sys.exit(1)

    print("Schema validation passed.")


if __name__ == "__main__":
    main()

File: submission/hrishikesh/src/main/java/com/kulu/sdet/ValidateSchema.java
package com.kulu.sdet;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.Map;
import java.util.Set;
import java.util.regex.Pattern;

/**
 * Standalone schema validator.
 *
 * <p>Rather than spinning up a database, this tool parses the Flyway migration SQL and verifies
 * that every table and column the test suite depends on is declared. This keeps the CI schema check
 * hermetic — no Docker, no JDBC — while still failing loudly if the schema drifts away from what
 * the tests assume.
 */
public final class ValidateSchema {

  private ValidateSchema() {}

  private static final Map<String, Set<String>> REQUIRED = new LinkedHashMap<>();

  static {
    REQUIRED.put("wallets", new LinkedHashSet<>(Set.of("id", "balance", "currency")));
    REQUIRED.put(
        "transfers",
        new LinkedHashSet<>(
            Set.of(
                "id",
                "source_wallet_id",
                "destination_wallet_id",
                "amount",
                "currency",
                "reference",
                "status",
                "idempotency_key",
                "created_at")));
    REQUIRED.put(
        "idempotency_keys",
        new LinkedHashSet<>(
            Set.of(
                "idempotency_key",
                "payload_hash",
                "transfer_id",
                "response_status",
                "response_body",
                "created_at")));
    REQUIRED.put(
        "transfer_events",
        new LinkedHashSet<>(Set.of("id", "transfer_id", "event_type", "payload", "created_at")));
    REQUIRED.put(
        "outbox_events",
        new LinkedHashSet<>(
            Set.of("id", "aggregate_id", "event_type", "payload", "published_at", "created_at")));
  }

  public static void main(String[] args) throws IOException {
    Path migrationsDir = Paths.get("src", "main", "resources", "db", "migration");
    if (!Files.isDirectory(migrationsDir)) {
      fail("Migrations directory not found: " + migrationsDir.toAbsolutePath());
    }

    StringBuilder combined = new StringBuilder();
    try (var stream = Files.list(migrationsDir)) {
      stream
          .filter(p -> p.getFileName().toString().endsWith(".sql"))
          .sorted()
          .forEach(
              p -> {
                try {
                  combined.append(Files.readString(p, StandardCharsets.UTF_8)).append('\n');
                } catch (IOException e) {
                  throw new RuntimeException(e);
                }
              });
    }
    String sql = combined.toString().toLowerCase();

    boolean ok = true;
    for (Map.Entry<String, Set<String>> e : REQUIRED.entrySet()) {
      String table = e.getKey();
      String tableBody = extractCreateTable(sql, table);
      if (tableBody == null) {
        System.err.println("MISSING TABLE: " + table);
        ok = false;
        continue;
      }
      for (String col : e.getValue()) {
        Pattern colPattern = Pattern.compile("(^|[,\\s(])" + Pattern.quote(col) + "\\s");
        if (!colPattern.matcher(tableBody).find()) {
          System.err.println("MISSING COLUMN: " + table + "." + col);
          ok = false;
        }
      }
    }

    // Guardrails for correctness the SUT relies on.
    if (!sql.contains("balance >= 0")) {
      System.err.println("MISSING CONSTRAINT: wallets.balance >= 0");
      ok = false;
    }
    if (!sql.contains("amount > 0")) {
      System.err.println("MISSING CONSTRAINT: transfers.amount > 0");
      ok = false;
    }
    if (!sql.contains("unique index ux_outbox_transfer_completed")) {
      System.err.println("MISSING UNIQUE INDEX: outbox_events(aggregate_id, event_type)");
      ok = false;
    }

    if (!ok) {
      fail("Schema validation failed. See errors above.");
    }
    System.out.println("Schema validation passed. Tables checked: " + REQUIRED.keySet());
  }

  private static String extractCreateTable(String sql, String table) {
    Pattern p =
        Pattern.compile(
            "create\\s+table\\s+" + Pattern.quote(table) + "\\s*\\((.*?)\\)\\s*;", Pattern.DOTALL);
    var m = p.matcher(sql);
    return m.find() ? m.group(1) : null;
  }

  private static void fail(String msg) {
    System.err.println(msg);
    System.exit(1);
  }
}
